test: Get rid of default wallet hacks

ryanofsky opened the pull request:
Changes:

- Get rid of setup_nodes (`-wallet`, `-nowallet`, `-disablewallet`) argument rewriting
- Get rid of hardcoded wallet `""` names and `-wallet=""` args

Motivation:

- Simplify test framework behavior so it's easier to write new tests without having arguments mangled by the framework
- Make tests more readable, replacing unexplained `""` string literals with `default_wallet_name` references
- Make it trivial to update default wallet name and wallet data filename for sqlite #19077 testing
- Stop relying on `-wallet` arguments to create wallets, so it is easy to change `-wallet` option in the future to only load existing wallets not create new ones (to avoid accidental wallet creation, and encourage use of wallet encryption and descriptor features)

ryanofsky committed:
test: Get rid of default wallet hacks

- Get rid of hardcoded wallet "" names and -wallet="" args
- Get rid of setup_nodes (-wallet, -nowallet, -disablewallet) argument rewriting

Motivation:

- Simplify test framework behavior so it's easier to write new tests without
  having arguments mangled by the framework
- Make tests more readable, replacing unexplained "" string literals with
  default_wallet_name references
- Make it trivial to update default wallet name and wallet data filename for
  sqlite #19077 testing
- Stop relying on -wallet arguments to create wallets, so it is easy to change
  -wallet option in the future to only load existing wallets not create new
  ones (to avoid accidental wallet creation, and encourage use of wallet
  encryption and descriptor features)

ryanofsky committed:
test, refactor: add default_wallet_name and wallet_data_filename variables

No changes in behavior

ryanofsky commented on test/functional/wallet_multiwallet.py:
```diff
@@ -108,7 +107,7 @@ def wallet_file(name):
         # should not initialize if wallet path can't be created
         exp_stderr = "boost::filesystem::create_directory:"
-        self.nodes[0].assert_start_raises_init_error(['-wallet=wallet.dat/bad'], exp_stderr, match=ErrorMatch.PARTIAL_REGEX)
+        self.nodes[0].assert_start_raises_init_error(['-wallet=w8/bad'], exp_stderr, match=ErrorMatch.PARTIAL_REGEX)
```
This change isn't strictly necessary, but using `w8` instead of `wallet.dat` here makes the test more robust so `default_wallet_filename` can be changed in the future. The top level wallet directory will always contain a `w8` file, but may or may not contain a `wallet.dat` file, depending on the `default_wallet_filename` value.

MarcoFalke commented:
Do all of the changes need to happen in one commit? Also, the build without wallet doesn't pass the tests

ryanofsky commented:
Thanks, I'll check the no wallet build. I don't think this is PR hard to review. It was very difficult to write, because every one change I made broke other things and required more changes. But in the end, the main change was just dropping -wallet="" arguments and tweaking things to work around the dropped arguments.

I would also expect if you're looking at one test and want to see how this PR affects the test, you'd rather look at one diff instead of multiple diffs. But I guess you have that option regardless by diffing across commits, so I'll see if there are sensible ways to have multiple commits.

ryanofsky commented:
Updated 9fbba7ca8712be85a72b8499c437c6259b171d03 -> c072c8585dde004db0aadbf4309fb97a952a6f29 ([`pr/defw.1`](https://github.com/ryanofsky/bitcoin/commits/pr/defw.1) -> [`pr/defw.2`](https://github.com/ryanofsky/bitcoin/commits/pr/defw.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/defw.1..pr/defw.2)) with suggestions from Marco (thanks!) splitting up commits and adding missing is_wallet_compiled check

achow101 commented:
I don't really like removing the `createwallet` override as it makes it more difficult to enable additional tests for descriptor wallets and lets us enable a test for descriptor wallets without having to modify every `createwallet` call in a test.

ryanofsky commented:
re: @achow101 https://github.com/bitcoin/bitcoin/pull/20034#issuecomment-700814066

> I don't really like removing the `createwallet` override as it makes it more difficult to enable additional tests for descriptor wallets and lets us enable a test for descriptor wallets without having to modify every `createwallet` call in a test.

Thanks! Dropped commit be968c406944a74550799d28200255436c880f4b as I didn't realize it was separable. The change does seem like a win to me for simplicity and making things explicit, so maybe I will try to revive it in some form. But there's no reason to include it here.

---

Updated c072c8585dde004db0aadbf4309fb97a952a6f29 -> 5338dfaef077b5c38e3a87d76f12068147a5db2d ([`pr/defw.2`](https://github.com/ryanofsky/bitcoin/commits/pr/defw.2) -> [`pr/defw.3`](https://github.com/ryanofsky/bitcoin/commits/pr/defw.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/defw.2..pr/defw.3)) just dropping commit be968c406944a74550799d28200255436c880f4b

ryanofsky commented on test/functional/tool_wallet.py:
```diff
@@ -213,7 +213,7 @@ def test_salvage(self):
         self.assert_tool_output('', '-wallet=salvage', 'salvage')
 
     def run_test(self):
-        self.wallet_path = os.path.join(self.nodes[0].datadir, self.chain, 'wallets', 'wallet.dat')
```
re: https://github.com/bitcoin/bitcoin/pull/20034#discussion_r496914396

> Should the `-wallet=wallet.dat` throughout this test use `self.wallet_data_filename` too then?

It effectively is. I think you are maybe only looking at the first commit, not the overall diff? The wallet.dat references are removed in the second commit, since more robust way to refer to wallets is by directory name not data file name. (The data file name used inside wallets is an implementation detail and referring to it is only supported in limited cases for backwards compatibility. Dropping references to the data file name lets the test work with all default_wallet_name values.)

ryanofsky commented on test/functional/wallet_multiwallet.py:
```diff
@@ -94,10 +94,10 @@ def wallet_file(name):
-        wallet_names = ['w1', 'w2', 'w3', 'w', 'sub/w5', os.path.join(self.options.tmpdir, 'extern/w6'), 'w7_symlink', 'w8', '']
-        extra_args = ['-wallet={}'.format(n) for n in wallet_names]
+        wallet_names = ['w1', 'w2', 'w3', 'w', 'sub/w5', os.path.join(self.options.tmpdir, 'extern/w6'), 'w7_symlink', 'w8', self.default_wallet_name]
+        extra_args = ['-nowallet'] + ['-wallet={}'.format(n) for n in wallet_names]
```
re: https://github.com/bitcoin/bitcoin/pull/20034#discussion_r496916277

> Why?

Specifying `-nowallet` keeps test behavior unchanged now that the load on startup option is used. Probably if you were writing this test from scratch, you would drop `-nowallet` and instead change the checks below. But this is the most minimal update for the existing test, it is also nice because it keeps the test more narrow and robust to outside changes.

MarcoFalke replied:
in the first commit:

The test passes without this change, so I am still wondering why this is needed.

Also, the commit doesn't seem to be exhaustive. The `wallet.dat/bad` should be replaced as well. Also, the usages in `tool_wallet.py`. See the diff I used for testing:

```diff
diff --git a/src/wallet/db.cpp b/src/wallet/db.cpp
index bd1d114730..89d7325e1c 100644
--- a/src/wallet/db.cpp
+++ b/src/wallet/db.cpp
@@ -20,6 +20,6 @@ void SplitWalletPath(const fs::path& wallet_path, fs::path& env_directory, std::
         // Normal case: Interpret wallet path as a directory path containing
         // data and log files.
         env_directory = wallet_path;
-        database_filename = "wallet.dat";
+        database_filename = "wallet_dat";
     }
 }
diff --git a/src/wallet/test/db_tests.cpp b/src/wallet/test/db_tests.cpp
index 8f0083cd2e..3489d48879 100644
--- a/src/wallet/test/db_tests.cpp
+++ b/src/wallet/test/db_tests.cpp
@@ -29,7 +29,7 @@ BOOST_AUTO_TEST_CASE(getwalletenv_file)
 
 BOOST_AUTO_TEST_CASE(getwalletenv_directory)
 {
-    std::string expected_name = "wallet.dat";
+    std::string expected_name = "wallet_dat";
     const fs::path datadir = GetDataDir();
 
     std::string filename;
diff --git a/src/wallet/walletutil.cpp b/src/wallet/walletutil.cpp
index e4c72aed98..3a16cb3e75 100644
--- a/src/wallet/walletutil.cpp
+++ b/src/wallet/walletutil.cpp
@@ -71,11 +71,11 @@ std::vector<fs::path> ListWalletDir()
         // This can be replaced by boost::filesystem::lexically_relative once boost is bumped to 1.60.
         const fs::path path = it->path().string().substr(offset);
 
-        if (it->status().type() == fs::directory_file && IsBerkeleyBtree(it->path() / "wallet.dat")) {
+        if (it->status().type() == fs::directory_file && IsBerkeleyBtree(it->path() / "wallet_dat")) {
             // Found a directory which contains wallet.dat btree file, add it as a wallet.
             paths.emplace_back(path);
         } else if (it.level() == 0 && it->symlink_status().type() == fs::regular_file && IsBerkeleyBtree(it->path())) {
-            if (it->path().filename() == "wallet.dat") {
+            if (it->path().filename() == "wallet_dat") {
                 // Found top-level wallet.dat btree file, add top level directory ""
                 // as a wallet.
                 paths.emplace_back();
diff --git a/test/functional/test_framework/test_framework.py b/test/functional/test_framework/test_framework.py
index 100be26c59..fa72420178 100755
--- a/test/functional/test_framework/test_framework.py
+++ b/test/functional/test_framework/test_framework.py
@@ -103,7 +103,7 @@ class BitcoinTestFramework(metaclass=BitcoinTestMetaClass):
         self.supports_cli = True
         self.bind_to_localhost_only = True
         self.default_wallet_name = ""
-        self.wallet_data_filename = "wallet.dat"
+        self.wallet_data_filename = "wallet_dat"
         self.set_test_params()
         self.parse_args()
         if self.options.timeout_factor == 0 :
diff --git a/test/functional/tool_wallet.py b/test/functional/tool_wallet.py
index ef68e224b6..8d3e7f8f62 100755
--- a/test/functional/tool_wallet.py
+++ b/test/functional/tool_wallet.py
@@ -73,7 +73,7 @@ class ToolWalletTest(BitcoinTestFramework):
         locked_dir = os.path.join(self.options.tmpdir, "node0", "regtest", "wallets")
         self.assert_raises_tool_error(
             'Error initializing wallet database environment "{}"!'.format(locked_dir),
-            '-wallet=wallet.dat',
+            '-wallet=wallet_dat',
             'info',
         )
         path = os.path.join(self.options.tmpdir, "node0", "regtest", "wallets", "nonexistent.dat")
@@ -104,7 +104,7 @@ class ToolWalletTest(BitcoinTestFramework):
             Transactions: 0
             Address Book: 3
         ''')
-        self.assert_tool_output(out, '-wallet=wallet.dat', 'info')
+        self.assert_tool_output(out, '-wallet=wallet_dat', 'info')
         timestamp_after = self.wallet_timestamp()
         self.log.debug('Wallet file timestamp after calling info: {}'.format(timestamp_after))
         self.log_wallet_timestamp_comparison(timestamp_before, timestamp_after)
@@ -143,7 +143,7 @@ class ToolWalletTest(BitcoinTestFramework):
             Transactions: 1
             Address Book: 3
         ''')
-        self.assert_tool_output(out, '-wallet=wallet.dat', 'info')
+        self.assert_tool_output(out, '-wallet=wallet_dat', 'info')
         shasum_after = self.wallet_shasum()
         timestamp_after = self.wallet_timestamp()
         self.log.debug('Wallet file timestamp after calling info: {}'.format(timestamp_after))
diff --git a/test/functional/wallet_dump.py b/test/functional/wallet_dump.py
index 09581d864b..5fdabf89c8 100755
--- a/test/functional/wallet_dump.py
+++ b/test/functional/wallet_dump.py
@@ -203,7 +203,7 @@ class WalletDumpTest(BitcoinTestFramework):
         assert result['ismine']
 
         self.log.info('Check that wallet is flushed')
-        with self.nodes[0].assert_debug_log(['Flushing wallet.dat'], timeout=20):
+        with self.nodes[0].assert_debug_log(['Flushing wallet_dat'], timeout=20):
             self.nodes[0].getnewaddress()
 
 
diff --git a/test/functional/wallet_multiwallet.py b/test/functional/wallet_multiwallet.py
index 7b975af065..71aa86abdb 100755
--- a/test/functional/wallet_multiwallet.py
+++ b/test/functional/wallet_multiwallet.py
@@ -95,7 +95,7 @@ class MultiWalletTest(BitcoinTestFramework):
         #   w8         - to verify existing wallet file is loaded correctly
         #   ''         - to verify default wallet file is created correctly
         wallet_names = ['w1', 'w2', 'w3', 'w', 'sub/w5', os.path.join(self.options.tmpdir, 'extern/w6'), 'w7_symlink', 'w8', self.default_wallet_name]
-        extra_args = ['-nowallet'] + ['-wallet={}'.format(n) for n in wallet_names]
+        extra_args = ['-wallet={}'.format(n) for n in wallet_names]
         self.start_node(0, extra_args)
         assert_equal(sorted(map(lambda w: w['name'], self.nodes[0].listwalletdir()['wallets'])), [self.default_wallet_name, os.path.join('sub', 'w5'), 'w', 'w1', 'w2', 'w3', 'w7', 'w7_symlink', 'w8'])
 
@@ -108,7 +108,7 @@ class MultiWalletTest(BitcoinTestFramework):
 
         # should not initialize if wallet path can't be created
         exp_stderr = "boost::filesystem::create_directory:"
-        self.nodes[0].assert_start_raises_init_error(['-wallet=wallet.dat/bad'], exp_stderr, match=ErrorMatch.PARTIAL_REGEX)
+        self.nodes[0].assert_start_raises_init_error(['-wallet={}/bad'.format(self.wallet_data_filename)], exp_stderr, match=ErrorMatch.PARTIAL_REGEX)
 
         self.nodes[0].assert_start_raises_init_error(['-walletdir=wallets'], 'Error: Specified -walletdir "wallets" does not exist')
         self.nodes[0].assert_start_raises_init_error(['-walletdir=wallets'], 'Error: Specified -walletdir "wallets" is a relative path', cwd=data_dir())

achow101 commented:
ACK 5338dfaef077b5c38e3a87d76f12068147a5db2d

MarcoFalke commented on test/functional/test_framework/test_framework.py:
```diff
@@ -102,6 +102,7 @@ def __init__(self):
         self.rpc_timeout = 60  # Wait for up to 60 seconds for the RPC server to respond
         self.supports_cli = True
         self.bind_to_localhost_only = True
+        self.wallet_names = None
```
in commit 5338dfaef077b5c38e3a87d76f12068147a5db2d:

Would be good to add a short comment here to say:

* Will be used to create wallets on setup
* Defaults to create a default wallet for each node
* List can be shorter than the number of nodes, or include None entries to skip wallet creation

MarcoFalke commented on test/functional/test_framework/test_framework.py:
```diff
@@ -396,13 +386,10 @@ def setup_nodes(self):
-                continue
-
+        wallet_names = [self.default_wallet_name] * len(self.nodes) if self.wallet_names is None else self.wallet_names
+        for wallet_name, n in zip(wallet_names, self.nodes):
```
Could add an assert to check self.num_nodes >= len(wallet_names)?

maflcko reviewed (commented):
Concept ACK 5338dfaef077b5c38e3a87d76f12068147a5db2d. Skimmed the two commits and left some coments.

ryanofsky reviewed (commented):
Thanks for the review!

Updated 5338dfaef077b5c38e3a87d76f12068147a5db2d -> c1585bca8dae01dee6a1dd8eadae2f8b100503df ([`pr/defw.3`](https://github.com/ryanofsky/bitcoin/commits/pr/defw.3) -> [`pr/defw.4`](https://github.com/ryanofsky/bitcoin/commits/pr/defw.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/defw.3..pr/defw.4)) with suggested changes

MarcoFalke commented:
ACK c1585bca8dae01dee6a1dd8eadae2f8b100503df, only effective change is adding documentation 🎵

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK c1585bca8dae01dee6a1dd8eadae2f8b100503df, only effective change is adding documentation 🎵
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhp/gv9HLJ3Y2W8Ok3zvRURR+0h8Qo5u1mOhbHmlS/HNUyM92HAL9OI7D+J+uZa
4ZlfbjiNaGgqEWcrbI1ptkPUy1VsBRCZKSmAMpJdQ9ahbY8UAQo/x1AyBYSnFdPD
MxSxfQ81YzwE8aCvezZdkBGx3t4f1EcBK2+2Q1e8EPjHhNOdZLfEAXoo0CIoBpB9
DzKEMh9hzRE5M3/ELloNLMhVy4XJ7HcH/GjoSPuxQggEKIDBluBiJDaCi2MnTV0C
vqj61OVr4ekci/YhpIvY049iW2XP3jp4qfYPHiInn8IK+swTlrSQ7dFjTfS10etr
wEmfNHWWlmvGQe1pUBIdabQ+FpwBC8CLg8NPavETm+mRx98EL2GwdPPCiPsU3oS+
VWttIL28ICzQZEpcDXcXij2XodcOANVlGifhmyxxxlyj2hVNU8cWp2Ep5fw1kcnH
eAn7wZztEJntWSKc7BhJW9EJ+rA+CX+OSMT7dBmqVVXUDP6XwaVlUoZ/sqWWPjcy
jTy++c1j
=XL4w
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `21bb7e58a106d518ab420a88873a29f5a4a1169244282ef31f53cec8fd2f3417  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010821bb7e58a106d518ab420a88873a29f5a4a1169244282ef31f53cec8fd2f3417f0105341ae36f5f1a93dad127b21d1374d2908fff01011e6a29c2cb49a3ade999f9c62f4254708f0203cae9f0e757d1ba2bd0064da63a4d612f28a664c8113ddfb560da175a61f1a2f08f1045f7741d9f008b4ffd91c837cd0ca0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010a158f29c4fb1fa4c99518f5396b7417708f0209b02a28087a08d383c0e72c9930389e003811a6ff441c78f87c6a6aed4d1d35508f1045f7741d8f008c36198091ee741560083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010e3277f1e828c2674bf9ec96c97b65b0308f020fb383a8a2c61e37e0f117b15df132e733ea29ba1221416a71e1b47cdb40506e608f1045f7741d8f008812736410382f6230083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010fa40fde691c1f28bd59f98325523ef5708f02082f6dd4e19258b9ee320cebf806284b8abacee74a48c2d3e068d37d86d36bde508f1045f7741d7f008ea71d0d1c820db1e0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

MarcoFalke commented:
Tested with 

```diff
diff --git a/src/wallet/db.cpp b/src/wallet/db.cpp
index bd1d114730..89d7325e1c 100644
--- a/src/wallet/db.cpp
+++ b/src/wallet/db.cpp
@@ -20,6 +20,6 @@ void SplitWalletPath(const fs::path& wallet_path, fs::path& env_directory, std::
         // Normal case: Interpret wallet path as a directory path containing
         // data and log files.
         env_directory = wallet_path;
-        database_filename = "wallet.dat";
+        database_filename = "wallet_dat";
     }
 }
diff --git a/src/wallet/test/db_tests.cpp b/src/wallet/test/db_tests.cpp
index 8f0083cd2e..3489d48879 100644
--- a/src/wallet/test/db_tests.cpp
+++ b/src/wallet/test/db_tests.cpp
@@ -29,7 +29,7 @@ BOOST_AUTO_TEST_CASE(getwalletenv_file)
 
 BOOST_AUTO_TEST_CASE(getwalletenv_directory)
 {
-    std::string expected_name = "wallet.dat";
+    std::string expected_name = "wallet_dat";
     const fs::path datadir = GetDataDir();
 
     std::string filename;
diff --git a/src/wallet/walletutil.cpp b/src/wallet/walletutil.cpp
index e4c72aed98..3a16cb3e75 100644
--- a/src/wallet/walletutil.cpp
+++ b/src/wallet/walletutil.cpp
@@ -71,11 +71,11 @@ std::vector<fs::path> ListWalletDir()
         // This can be replaced by boost::filesystem::lexically_relative once boost is bumped to 1.60.
         const fs::path path = it->path().string().substr(offset);
 
-        if (it->status().type() == fs::directory_file && IsBerkeleyBtree(it->path() / "wallet.dat")) {
+        if (it->status().type() == fs::directory_file && IsBerkeleyBtree(it->path() / "wallet_dat")) {
             // Found a directory which contains wallet.dat btree file, add it as a wallet.
             paths.emplace_back(path);
         } else if (it.level() == 0 && it->symlink_status().type() == fs::regular_file && IsBerkeleyBtree(it->path())) {
-            if (it->path().filename() == "wallet.dat") {
+            if (it->path().filename() == "wallet_dat") {
                 // Found top-level wallet.dat btree file, add top level directory ""
                 // as a wallet.
                 paths.emplace_back();
diff --git a/test/functional/test_framework/test_framework.py b/test/functional/test_framework/test_framework.py
index 2824d80434..6ab813ff1c 100755
--- a/test/functional/test_framework/test_framework.py
+++ b/test/functional/test_framework/test_framework.py
@@ -104,7 +104,7 @@ class BitcoinTestFramework(metaclass=BitcoinTestMetaClass):
         self.bind_to_localhost_only = True
         self.parse_args()
         self.default_wallet_name = ""
-        self.wallet_data_filename = "wallet.dat"
+        self.wallet_data_filename = "wallet_dat"
         # Optional list of wallet names that can be set in set_test_params to
         # create and import keys to. If unset, default is len(nodes) *
         # [default_wallet_name]. If wallet names are None, wallet creation is
diff --git a/test/functional/wallet_dump.py b/test/functional/wallet_dump.py
index 09581d864b..5fdabf89c8 100755
--- a/test/functional/wallet_dump.py
+++ b/test/functional/wallet_dump.py
@@ -203,7 +203,7 @@ class WalletDumpTest(BitcoinTestFramework):
         assert result['ismine']
 
         self.log.info('Check that wallet is flushed')
-        with self.nodes[0].assert_debug_log(['Flushing wallet.dat'], timeout=20):
+        with self.nodes[0].assert_debug_log(['Flushing wallet_dat'], timeout=20):
             self.nodes[0].getnewaddress()
